File: couple-carrier-risk/r/tests/testthat/test-couple_risk_calculator.R
ar_row <- function(name, cf, couple = NULL) {
  list(
    disease_name = name,
    inheritance_pattern = INHERITANCE_AR,
    carrier_frequency = cf,
    couple_offspring_risk = couple
  )
}

xlr_row <- function(name, cf, couple = NULL) {
  list(
    disease_name = name,
    inheritance_pattern = INHERITANCE_XLR,
    carrier_frequency = cf,
    couple_offspring_risk = couple
  )
}

risks_named_by_disease <- function(...) {
  entries <- list(...)
  stats::setNames(entries, vapply(entries, function(e) e$disease_name, character(1L)))
}

test_that("parse_population_risks indexes by disease_name", {
  indexed <- parse_population_risks(load_json_fixture("sample-population-risks"))

  expect_setequal(
    names(indexed),
    c("Sickle cell anaemia", "Cystic fibrosis", "Duchenne muscular dystrophy")
  )
  expect_equal(indexed[["Sickle cell anaemia"]]$inheritance_pattern, INHERITANCE_AR)
})

test_that("parse_population_risks raises without a 'risks' list", {
  expect_error(parse_population_risks(list()), class = "response_schema_error")
})

test_that("AR cross-partner risk is cf_a * cf_b / 4", {
  partner_a <- list(
    biological_sex = SEX_MALE,
    risks = risks_named_by_disease(ar_row("Sickle cell", 0.08, couple = 0.0016))
  )
  partner_b <- list(
    biological_sex = SEX_FEMALE,
    risks = risks_named_by_disease(ar_row("Sickle cell", 0.05, couple = 0.000625))
  )

  rows <- build_couple_rows(partner_a, partner_b)

  expect_length(rows, 1L)
  expect_equal(rows[[1L]]$carrier_frequency_a, 0.08)
  expect_equal(rows[[1L]]$carrier_frequency_b, 0.05)
  expect_equal(rows[[1L]]$cross_partner_offspring_risk, 0.08 * 0.05 / 4)
})

test_that("AR with one missing cf yields no cross-partner risk", {
  partner_a <- list(
    biological_sex = SEX_MALE,
    risks = risks_named_by_disease(ar_row("X", 0.05, couple = 0.0006))
  )
  partner_b <- list(biological_sex = SEX_FEMALE, risks = list())

  rows <- build_couple_rows(partner_a, partner_b)

  expect_null(rows[[1L]]$cross_partner_offspring_risk)
  expect_null(rows[[1L]]$carrier_frequency_b)
})

test_that("XLR uses only the female partner's carrier frequency", {
  partner_a <- list(
    biological_sex = SEX_MALE,
    risks = risks_named_by_disease(xlr_row("DMD", 0, couple = 0))
  )
  partner_b <- list(
    biological_sex = SEX_FEMALE,
    risks = risks_named_by_disease(xlr_row("DMD", 0.002, couple = 0.0005))
  )

  rows <- build_couple_rows(partner_a, partner_b)

  expect_equal(rows[[1L]]$cross_partner_offspring_risk, 0.002 / 4)
})

test_that("XLR with two male partners yields no cross-partner risk", {
  partner_a <- list(
    biological_sex = SEX_MALE,
    risks = risks_named_by_disease(xlr_row("DMD", 0.002))
  )
  partner_b <- list(
    biological_sex = SEX_MALE,
    risks = risks_named_by_disease(xlr_row("DMD", 0.002))
  )

  rows <- build_couple_rows(partner_a, partner_b)
  expect_null(rows[[1L]]$cross_partner_offspring_risk)
})

test_that("row order is partner-A then B-only entries", {
  partner_a <- list(
    biological_sex = SEX_MALE,
    risks = risks_named_by_disease(
      ar_row("A-only", 0.01),
      ar_row("Shared", 0.02)
    )
  )
  partner_b <- list(
    biological_sex = SEX_FEMALE,
    risks = risks_named_by_disease(
      ar_row("Shared", 0.03),
      ar_row("B-only", 0.04)
    )
  )

  rows <- build_couple_rows(partner_a, partner_b)

  expect_equal(
    vapply(rows, function(row) row$disease_name, character(1L)),
    c("A-only", "Shared", "B-only")
  )
})

test_that("non-numeric carrier frequency raises response_schema_error", {
  bad <- list(
    disease_name = "Bad",
    inheritance_pattern = INHERITANCE_AR,
    carrier_frequency = "oops"
  )
  partner_a <- list(
    biological_sex = SEX_MALE,
    risks = stats::setNames(list(bad), "Bad")
  )
  partner_b <- list(biological_sex = SEX_FEMALE, risks = list())

  expect_error(
    build_couple_rows(partner_a, partner_b),
    class = "response_schema_error"
  )
})

File: couple-carrier-risk/python/tests/test_couple_risk_calculator.py
from typing import Any

import pytest

from couple_carrier_risk.couple_risk_calculator import (
    INHERITANCE_AR,
    INHERITANCE_XLR,
    PartnerRisks,
    ResponseSchemaError,
    build_couple_rows,
    parse_population_risks,
)
from couple_carrier_risk.genome_file import BiologicalSex
from tests.fixtures_loader import load_json_fixture


def _risks(*entries: dict[str, Any]) -> dict[str, dict[str, Any]]:
    return {entry["disease_name"]: entry for entry in entries}


def _ar(name: str, cf: float | None, *, couple: float | None = None) -> dict[str, Any]:
    return {
        "disease_name": name,
        "inheritance_pattern": INHERITANCE_AR,
        "carrier_frequency": cf,
        "couple_offspring_risk": couple,
    }


def _xlr(name: str, cf: float | None, *, couple: float | None = None) -> dict[str, Any]:
    return {
        "disease_name": name,
        "inheritance_pattern": INHERITANCE_XLR,
        "carrier_frequency": cf,
        "couple_offspring_risk": couple,
    }


def test_parse_population_risks_indexes_by_disease_name() -> None:
    indexed = parse_population_risks(load_json_fixture("sample-population-risks"))

    assert set(indexed) == {
        "Sickle cell anaemia",
        "Cystic fibrosis",
        "Duchenne muscular dystrophy",
    }
    assert indexed["Sickle cell anaemia"]["inheritance_pattern"] == INHERITANCE_AR


def test_parse_population_risks_requires_risks_list() -> None:
    with pytest.raises(ResponseSchemaError):
        parse_population_risks({})


def test_ar_cross_partner_risk_is_cf_a_times_cf_b_over_four() -> None:
    partner_a = PartnerRisks(
        biological_sex=BiologicalSex.MALE,
        risks=_risks(_ar("Sickle cell", 0.08, couple=0.0016)),
    )
    partner_b = PartnerRisks(
        biological_sex=BiologicalSex.FEMALE,
        risks=_risks(_ar("Sickle cell", 0.05, couple=0.000625)),
    )

    rows = build_couple_rows(partner_a, partner_b)

    assert len(rows) == 1
    row = rows[0]
    assert row.carrier_frequency_a == 0.08
    assert row.carrier_frequency_b == 0.05
    assert row.cross_partner_offspring_risk == pytest.approx(0.08 * 0.05 / 4)


def test_ar_with_one_missing_cf_yields_no_cross_partner_risk() -> None:
    partner_a = PartnerRisks(BiologicalSex.MALE, _risks(_ar("X", 0.05, couple=0.0006)))
    partner_b = PartnerRisks(BiologicalSex.FEMALE, {})

    rows = build_couple_rows(partner_a, partner_b)

    assert rows[0].cross_partner_offspring_risk is None
    assert rows[0].carrier_frequency_b is None


def test_xlr_uses_only_the_female_partners_carrier_frequency() -> None:
    partner_a = PartnerRisks(BiologicalSex.MALE, _risks(_xlr("DMD", 0.0, couple=0.0)))
    partner_b = PartnerRisks(
        biological_sex=BiologicalSex.FEMALE,
        risks=_risks(_xlr("DMD", 0.002, couple=0.0005)),
    )

    rows = build_couple_rows(partner_a, partner_b)

    assert rows[0].cross_partner_offspring_risk == pytest.approx(0.002 / 4)


def test_xlr_with_no_female_partner_yields_no_cross_partner_risk() -> None:
    partner_a = PartnerRisks(BiologicalSex.MALE, _risks(_xlr("DMD", 0.002)))
    partner_b = PartnerRisks(BiologicalSex.MALE, _risks(_xlr("DMD", 0.002)))

    rows = build_couple_rows(partner_a, partner_b)

    assert rows[0].cross_partner_offspring_risk is None


def test_rows_union_preserves_partner_a_order_then_b_only_entries() -> None:
    partner_a = PartnerRisks(
        biological_sex=BiologicalSex.MALE,
        risks=_risks(_ar("A-only", 0.01), _ar("Shared", 0.02)),
    )
    partner_b = PartnerRisks(
        biological_sex=BiologicalSex.FEMALE,
        risks=_risks(_ar("Shared", 0.03), _ar("B-only", 0.04)),
    )

    rows = build_couple_rows(partner_a, partner_b)

    assert [row.disease_name for row in rows] == ["A-only", "Shared", "B-only"]


def test_non_numeric_carrier_frequency_raises_schema_error() -> None:
    partner_a = PartnerRisks(
        biological_sex=BiologicalSex.MALE,
        risks=_risks({
            "disease_name": "Bad",
            "inheritance_pattern": INHERITANCE_AR,
            "carrier_frequency": "oops",
        }),
    )
    partner_b = PartnerRisks(BiologicalSex.FEMALE, {})

    with pytest.raises(ResponseSchemaError):
        build_couple_rows(partner_a, partner_b)
